db/asset.go

```diff
@@ -0,0 +1,91 @@
+package db
+
+import (
+	"context"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+	"time"
+
+	"github.com/trustwallet/blockatlas/db/models"
+)
+
+func (i *Instance) AddNewAssets(assets []models.Asset, ctx context.Context) error {
+	db := i.Gorm.WithContext(ctx)
+	if len(assets) == 0 {
+		return nil
+	}
+	uniqueAssets := getUniqueAssets(assets)
+	existingAssets, err := i.GetAssetsByIDs(models.AssetIDs(uniqueAssets), ctx)
+	if err != nil {
+		return err
+	}
+	if len(existingAssets) == 0 {
+		return db.Clauses(clause.OnConflict{DoNothing: true}).Create(&uniqueAssets).Error
+	}
+	allAssetsMap := make(map[string]models.Asset)
+	for _, ua := range uniqueAssets {
+		allAssetsMap[ua.Asset] = ua
+	}
+	existingAssetsMap := make(map[string]models.Asset)
+	for _, ea := range existingAssets {
+		existingAssetsMap[ea.Asset] = ea
+	}
+	var newAssets []models.Asset
+	for k, v := range allAssetsMap {
+		_, ok := existingAssetsMap[k]
+		if !ok && v.Asset != "" {
+			newAssets = append(newAssets, v)
+		}
+	}
+	if len(newAssets) == 0 {
+		return nil
+	}
+
+	assetsBatch := assetsBatch(newAssets, batchCount)
+
+	return db.Transaction(func(tx *gorm.DB) error {
+		for _, na := range assetsBatch {
+			err := tx.Clauses(clause.OnConflict{DoNothing: true}).Create(&na).Error
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func assetsBatch(values []models.Asset, sizeUint uint) [][]models.Asset {
```

Comment:
Can you move this into local lib? I think this is used somewhere else already

Reply:
will use memory cache

Reply:
https://github.com/trustwallet/blockatlas/issues/1251